### PR TITLE
Token parser

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

File: Applications/Spire/Source/StyleParser/DataTypes/TokenParser.cpp [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  const auto PUNCTUATIONS =
+    std::unordered_map<std::string, Punctuation::PunctuationType>{
+      {":", Punctuation::PunctuationType::COLON},
+      {",", Punctuation::PunctuationType::COMMA},
+      {".", Punctuation::PunctuationType::PERIOD},
+      {";", Punctuation::PunctuationType::SEMI_COLON}};
+
+  const auto LITERAL_REGEX_TEXT_LIST =
+    std::unordered_map<LITERAL_TYPE, std::string>{
+      {LITERAL_TYPE::STRING, R"(\".*\")"},
```
Comment (kranar) [inline Applications/Spire/Source/StyleParser/DataTypes/TokenParser.cpp:68]: As discussed in our meeting, you can't use a regex to match a string literal.

**Review by kranar — CHANGES_REQUESTED** [review]

File: Applications/Spire/Include/Spire/StyleParser/DataTypes/Token.hpp [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+        /** The closing round bracket. */
+        CLOSE_ROUND
+      };
+
+      /**
+       * Constructs a Bracket.
+       * @param type The bracket type.
+       */
+      explicit Bracket(BracketType type);
```
[the remark this answers is not in the document]
Comment (shaobinspire) [reply Applications/Spire/Include/Spire/StyleParser/DataTypes/Token.hpp:40]: Based on the description of the task TokenParser, the brackets can be represented by a class called Bracket and similar classes will exist for each type of token, so I created these classes.

[the remark this answers is not in the document]
Comment (kranar) [reply Applications/Spire/Include/Spire/StyleParser/DataTypes/Token.hpp:40]: The API is overly complicated. Looking over the test cases it's very hard to use this API to perform simple tasks.

[the remark this answers is not in the document]
Comment (shaobinspire) [reply Applications/Spire/Include/Spire/StyleParser/DataTypes/Token.hpp:40]: I see.  I will remove these classes wrapping `XXXType` and keep those enumeration classes. For example, the class `Bracket` will be removed, and the enum class `BracketType` will remain but be renamed to `Bracket`.
```
  /** Specifies the bracket type. */
  enum class Bracket {

    /** The opening curly bracket. */
    OPEN_CURLY,

    /** The closing curly bracket. */
    CLOSE_CURLY,

    /** The opening square bracket. */
    OPEN_SQUARE,

    /** The closing square bracket. */
    CLOSE_SQUARE,

    /** The opening round bracket. */
    OPEN_ROUND,

    /** The closing round bracket. */
    CLOSE_ROUND
  };
```

[the remark this answers is not in the document]
Comment (kranar) [reply Applications/Spire/Include/Spire/StyleParser/DataTypes/Token.hpp:40]: That's a step in the right direction for sure. Feel free to review your API based on your use cases and test cases and then revise/update the API along with the implementation. You can resubmit this for tomorrow.

**Review by kranar — COMMENTED** [review]
There are a lot of issues with this both in terms of the API and the implementation, but there isn't much time left for this project.

I can't merge this into master in its current state, but you are welcome to work on your next task for the selector parser and then submit a pull request to merge it into this branch. In the process of working on the selector parser you will almost certainly need to make changes to the API and the implementation.

**Pull request closed without merging**